src/index.js

```diff
@@ -3,247 +3,267 @@
+            break;
+          }
+          regexStr = condition[key];
+          condition[key] = (astNode, comment) => new RegExp(regexStr).test(comment.value);
+          break;
+        default:
+          regex = condition[key];
+          condition[key] = (astNode, comment) => (regex.test(comment.value));
+      }
+    });
+
+    // Redefine the comments function to extract and preserve
+    // comments according to the two conditions
+    return (astNode, comment) => {
+      if (condition.extract(astNode, comment)) {
+        extractedComments.push(
+          comment.type === 'comment2' ? `/*${comment.value}*/` : `//${comment.value}`,
+        );
+      }
+      return condition.preserve(astNode, comment);
+    };
+  }
 
+  apply(compiler) {
     const requestShortener = new RequestShortener(compiler.context);
-    compiler.plugin('compilation', (compilation) => {
-      if (options.sourceMap) {
-        compilation.plugin('build-module', (module) => {
+    compiler.plugin('compilation', (compilationArg) => {
+      const compilation = compilationArg;
+
+      if (this.options.sourceMap) {
+        compilation.plugin('build-module', (moduleArg) => {
           // to get detailed location info about errors
-          module.useSourceMap = true;
+          const moduleVar = moduleArg;
+          moduleVar.useSourceMap = true;
         });
       }
+
       compilation.plugin('optimize-chunk-assets', (chunks, callback) => {
-        const files = [];
-        // eslint-disable-next-line prefer-spread
-        chunks.forEach(chunk => files.push.apply(files, chunk.files));
-        // eslint-disable-next-line prefer-spread
-        files.push.apply(files, compilation.additionalChunkAssets);
-        const filteredFiles = files.filter(ModuleFilenameHelpers.matchObject.bind(undefined, options));
         const uglifiedAssets = new WeakSet();
-        filteredFiles.forEach((file) => {
-          const oldWarnFunction = uglify.AST_Node.warn_function;
-          const warnings = [];
-          let sourceMap;
-          try {
+        chunks.reduce((acc, chunk) => acc.concat(chunk.files || []), [])
+          .concat(compilation.additionalChunkAssets || [])
+          .filter(ModuleFilenameHelpers.matchObject.bind(null, this.options))
+          .forEach((file) => {
+            // Copy uglify options
+            const uglifyOptions = UglifyJsPlugin.buildDefaultUglifyOptions(this.uglifyOptions);
+            let sourceMap;
             const asset = compilation.assets[file];
             if (uglifiedAssets.has(asset)) {
               return;
             }
-            let input;
-            let inputSourceMap;
-            if (options.sourceMap) {
-              if (asset.sourceAndMap) {
-                const sourceAndMap = asset.sourceAndMap();
-                inputSourceMap = sourceAndMap.map;
-                input = sourceAndMap.source;
+
+            try {
```

Comment:
@mikesherov I made this, if you want to have a look, but I'm not so convinced. https://gist.github.com/Hartorn/14cba0e11382c38a91c6ac1b0c0cc9c1

The question is :  if the user give wrong banner, warningsFilter or extractComments functions, should it crash, or go on the compilation ?

I'm not sure it's going to be a big win, since Uglify is using inside a try-catch : https://github.com/mishoo/UglifyJS2/blob/master/lib/minify.js#L49